### PR TITLE
WIP - adding functionality to resample Neurovault fetched images

### PR DESCRIPTION
**Pull request by nicofarr**
Related to issue #1553 

In order to deal with both resampled and non-resampled version, we need to implement a mechanism that checks the presence or absence of the resampled or original files. 

WIP - currently I'm not happy with the way it's implemented, because it ONLY works if you use the `mode=overwrite` in the fetch_neurovault functions. 

This is because I implemented the caching mechanism at the lowest level (download function), but the check for downloaded images seems to happen earlier.

### CODE REVIEW

File: nilearn/datasets/neurovault.py [hunk 1]
```diff
@@ -1507,11 +1513,71 @@ def _download_image_nii_file(image_info, collection, download_params):
...
+    #       - if not, check whether the non-resampled image is on disk
+    #           - if yes : use this image and resample it , keep the non resampled version
+    #           - if not : download the image as a tempfile, resample it, remove the tempfile
+
+
+    if download_params['resample']:
+
+
+
+        resampled_image_file_name = 'image_{0}_resampled.nii.gz'.format(image_id)
```
Comment (bthirion) [inline nilearn/datasets/neurovault.py]: Don't skip 3 lines.

Comment (bthirion) [inline nilearn/datasets/neurovault.py]: Why not `'{0}_resampled.nii.gz'.format(image_id)`

File: nilearn/datasets/neurovault.py [hunk 2]
```diff
@@ -44,6 +45,11 @@
 _DEFAULT_BATCH_SIZE = 100
 _DEFAULT_MAX_IMAGES = 100
 
+AFFINE_STD = np.array([[3., 0., 0., -90.],
```
Comment (bthirion) [inline nilearn/datasets/neurovault.py]: `STD_AFFINE` maybe.

Comment (bthirion) [inline nilearn/datasets/neurovault.py]: 2 things:
* I think that the standard affine you propose corresponds to a certain standard shape (e.g. to have approximate symmetry in x). I would add it.
* What if the use wants to resample differntly, e.g. in 2- or 4-mm resolution ? Could he/she provide an affine ?

File: nilearn/datasets/neurovault.py [hunk 3]
```diff
@@ -2116,6 +2182,7 @@ def _read_download_params(
     wanted_image_ids=None, max_images=None,
     max_consecutive_fails=_MAX_CONSECUTIVE_FAILS,
     max_fails_in_collection=_MAX_FAILS_IN_COLLECTION,
+    resample = False,
```
Comment (bthirion) [inline nilearn/datasets/neurovault.py]: `resample=False`

File: nilearn/datasets/neurovault.py [hunk 4]
```diff
@@ -2258,7 +2326,7 @@ def _fetch_neurovault_implementation(
     max_images=_DEFAULT_MAX_IMAGES, collection_terms=basic_collection_terms(),
     collection_filter=_empty_filter, image_terms=basic_image_terms(),
     image_filter=_empty_filter, collection_ids=None, image_ids=None,
-    mode='download_new', data_dir=None, fetch_neurosynth_words=False,
+    mode='download_new', data_dir=None, fetch_neurosynth_words=False,resample=False,
```
Comment (bthirion) [inline nilearn/datasets/neurovault.py]: a space after comma.

File: nilearn/datasets/neurovault.py [hunk 5]
```diff
@@ -2274,7 +2342,7 @@ def _fetch_neurovault_implementation(
         collection_terms=collection_terms,
         collection_filter=collection_filter, image_terms=image_terms,
         image_filter=image_filter, wanted_collection_ids=collection_ids,
-        wanted_image_ids=image_ids, max_images=max_images, verbose=verbose,
+        wanted_image_ids=image_ids, max_images=max_images, resample = resample,verbose=verbose,
```
Comment (bthirion) [inline nilearn/datasets/neurovault.py]: `resample=resample`

File: nilearn/datasets/neurovault.py [hunk 6]
```diff
@@ -2533,6 +2604,9 @@ def fetch_neurovault_ids(
     fetch_neurosynth_words : bool, optional (default=False)
         whether to collect words from Neurosynth.
 
+    resample : bool, optional (default=False)
+        Resamples downloaded images to a 4x4x4 grid before saving them, to save disk space.
```
Comment (bthirion) [inline nilearn/datasets/neurovault.py]: Sorry, what does '4x4x4 grid' mean ?

File: nilearn/datasets/neurovault.py [hunk 7]
```diff
@@ -2598,5 +2672,5 @@ def fetch_neurovault_ids(
         mode=mode,
         collection_ids=collection_ids, image_ids=image_ids,
         data_dir=data_dir,
-        fetch_neurosynth_words=fetch_neurosynth_words,
+        fetch_neurosynth_words=fetch_neurosynth_words,resample=resample,
```
Comment (bthirion) [inline nilearn/datasets/neurovault.py]: space after comma

**Comment by jeromedockes** [comment]
Thanks for working on this!

yes you're right, ``_download_image`` is in charge of actually downloading images and is not involved when they are already on the user's disk. ``_scroll`` starts by going over local images, yielding those that match the filter, then goes over the server's images. ``_update_image`` is called once for each local image and returns the image's metadata (it does nothing when user has no write permissions), and ``_download_image`` is called to download each missing image and returns the image's metadata. Maybe you could have each of these functions  ``return _your_resampling_cached_function(image_info)`` instead of ``return image_info``

Note that when downloading the image, if it is resampled, you will discard the original one to save space, but when getting it from the disk, you probably don't want to delete the original. We also need to decide what to do when the user doesn't have write permissions.

Maybe the metadata could contain several keys, e.g. 'absolute_path', 'absolute_path_resampled_3_mm', etc, and the returned Bunch.images list, which contains the niimg-like objects (nii file paths), would be built by selecting from those the paths to the appropriately resampled images.

Finally, I see that resample is on/off, and always at 3mm. Would it be useful to let the user choose the target voxel size?

**Comment by GaelVaroquaux** [comment]
> Finally, I see that resample is on/off, and always at 3mm. Would it be useful
> to let the user choose the target voxel size?

We choose not to do that to decrease the amount of different options /
pipeline through which the data can go: it is bad for caching, and gives
researchers degrees of freedom (bad for science).

**Comment by jeromedockes** [comment]
> We choose not to do that to decrease the amount of different options / pipeline through which the data can go: it is bad for caching, and gives researchers degrees of freedom (bad for science).

I understand, those are two good reasons.

File: nilearn/datasets/neurovault.py (hunk 1 above)
Comment (jeromedockes) [reply nilearn/datasets/neurovault.py]: I imagine to be similar to the original image file names, which are stored as `'image_{}.nii.gz'.format(image_id)`, where `image_id` is an `int`: image_123.nii.gz, image_123_resampled.nii.gz

**Comment by kchawla-pi** [comment]
Is this alive? If not, please close it by this weekend.

**Comment by nicofarr** [comment]
Hi,

I'll resume work on this during the Xmas holiday break  ! If not, I'll
close it.

Best,

Nico

**Comment by nicofarr** [comment]
Resuming this now, with the objective of having it merge during OHBM in Rome ! 
Will (most) probably have questions in the coming days / weeks to get this done.

**Comment by kchawla-pi** [comment]
Hey thanks for restarting this!
You should pull from master before making any more changes.

**Comment by nicofarr** [comment]
Reviving this thread. 

Here is a proposal for the intended behavior : 

if mode==overwrite, the images are downloaded and resampled only if resample=True

if overwrite is False, then it depends whether the images were already downloaded or not. 

- if the images are not on the disk, they are downloaded, resampled, and only the resampled version is kept. 

- if the images are already on the disk, they are resampled. Original images are kept, just in case other users acess the nilearn_data directory with the original images. 

There is a possibilit that both versions, resampled and original, are present on the disk, which could be an issue for large neurovault collections.

**Comment by bthirion** [comment]
>     * if the images are already on the disk, they are resampled. Original images are kept, just in case other users acess the nilearn_data directory with the original images.

But then you would have to rename them ? Won't this make the things messy in in-homogeneous on disk ?

**Comment by jeromedockes** [comment]
> But then you would have to rename them ? Won't this make the things messy in in-homogeneous on disk ?

yes with this option there is a possibility that some users will end up with both `image_123.nii.gz` and `image_123_resampled.nii.gz`. but i'm not sure there is a way to avoid this, since we don't want to erase anything from the user's disk without them explicitly asking for it. hence, if the original image is on disk and the user asks for the resampled image, we have to compute it and store it without erasing the original. because:
- if they ask for resampled, they will expect all images to have exactly the same affine and fov so returning the original is not an option
- we want to return a filename, not an in-memory image, in case we are dealing with a large collection.

in the same way, if the resampled is on disk and the user asks for the original we will download the original and leave the resampled one so the user will have both.

we can possibly issue a warning when the user ends up storing the same image twice.

this is less than ideal but it is what we thought of so far, suggestions are very welcome

**Comment by nicofarr** [comment]
I'll continue contributing to this during the nilearn coding sprint of Nilearn DevDays 2020!

**Comment by jeromedockes** [comment]
This PR has 3.000 commits and is touching 400 files :D could you rebase on master or start a fresh one? thanks

**Comment by nicofarr** [comment]
sorry, messed up when integrating the changes from master. I'll do as suggested. ;)

**Comment by jeromedockes** [comment]
great, thanks!

**Comment by nicofarr** [comment]
now writing the tests, will ping here once ready.

**Comment by nicofarr** [comment]
OK so here is a first version. As previously discussed, I implemented the following behavior : 

- if nothing is on disk and user asks for resampled version : the original version is downloaded in a temp file and resampled immediately, temp file is not kept, data['images'] contains a bunch with the resampled images
- If the original version is already on disk, no need for download, only resample. This is taken care by the _scroll_local function that now tests for the need to resample or not
- If a user asks for the original version (resample = False) but the resampled version is present on disk, then the data is downloaded again. Both files are on disk, but we don't want to remove anything as discussed above. 

I have proposed a test function that should check all those things. I didn't find how to check that download DOES happen or does NOT happen, @jeromedockes do you think this is feaasible ? 

The test function that I propose is a bit long (and downloads a neurovault collection with 3 images several times), should it be broken down ?

**Comment by jeromedockes** [comment]
the way to do this would be to mock the fetch_file  or simple_download function

File: nilearn/datasets/neurovault.py [hunk 8]
```diff
@@ -2279,7 +2338,7 @@ def _fetch_neurovault_implementation(
         collection_terms=collection_terms,
         collection_filter=collection_filter, image_terms=image_terms,
         image_filter=image_filter, wanted_collection_ids=collection_ids,
-        wanted_image_ids=image_ids, max_images=max_images, verbose=verbose,
+        wanted_image_ids=image_ids, max_images=max_images, resample=resample,verbose=verbose,
```
Comment (bthirion) [inline nilearn/datasets/neurovault.py:2341]:
```suggestion
        wanted_image_ids=image_ids, max_images=max_images, resample=resample, verbose=verbose,
```

File: nilearn/datasets/neurovault.py [hunk 9]
```diff
@@ -2361,6 +2420,9 @@ def fetch_neurovault(
         counts and add it to the result. Also add to the result a
         vocabulary list. See ``sklearn.CountVectorizer`` for more info.
 
+    resample : bool, optional (default=False)
+        Resamples downloaded images to a 4x4x4 grid before saving them, to save disk space.
```
Comment (bthirion) [inline nilearn/datasets/neurovault.py:2424]: '4x4x4 grid' : WDYM ?
The output resolution is actually 3mm x 3mm x 3mm

File: nilearn/datasets/tests/test_neurovault.py [hunk 10]
```diff
@@ -622,3 +622,74 @@ def test_fetch_neurovault_ids():
...
+
+        # Check the expected size of the dataset
+        assert (len(data['images_meta'])) == 3
+        
+        # Check that the resampled version is here
+        assert np.all([os.path.isfile(im_meta['resampled_absolute_path']) for im_meta in data['images_meta']])
+
+        # Load images that are fetched and check the affines
+        affines = [load_img(cur_im).affine for cur_im in data['images']]
+        assert np.all([np.all(affine==neurovault.STD_AFFINE) for affine in affines])
```
Comment (bthirion) [inline nilearn/datasets/tests/test_neurovault.py:641]:
```suggestion
        assert np.all([np.all(affine == neurovault.STD_AFFINE) for affine in affines])
```

File: nilearn/datasets/tests/test_neurovault.py [hunk 11]
```diff
@@ -622,3 +622,74 @@ def test_fetch_neurovault_ids():
...
+
+        # Check that the original version is NOT here
+        assert not np.any([os.path.isfile(im_meta['absolute_path']) for im_meta in data['images_meta']])
+
+        # Ask to download the non-resampled version. This should trigger download
+        # (TODO - How can we test systematically that download does happen ?  )
+        data_orig = neurovault.fetch_neurovault_ids(collection_ids=[4666], data_dir=data_dir,resample=False)
```
Comment (bthirion) [inline nilearn/datasets/tests/test_neurovault.py:648]:
```suggestion
        data_orig = neurovault.fetch_neurovault_ids(collection_ids=[4666], data_dir=data_dir, resample=False)
```

File: nilearn/datasets/tests/test_neurovault.py [hunk 12]
```diff
@@ -622,3 +622,74 @@ def test_fetch_neurovault_ids():
...
+
+        # Check that the original version is now here
+        assert np.all([os.path.isfile(im_meta['absolute_path']) for im_meta in data_orig['images_meta']])
+
+        # Check that the affines of the original version do not correspond to the resampled one
+        affines_orig = [load_img(cur_im).affine for cur_im in data_orig['images']]
+        assert not np.any([np.all(affine==neurovault.STD_AFFINE) for affine in affines_orig])
```
Comment (bthirion) [inline nilearn/datasets/tests/test_neurovault.py:655]:
```suggestion
        assert not np.any([np.all(affine == neurovault.STD_AFFINE) for affine in affines_orig])
```

File: nilearn/datasets/tests/test_neurovault.py [hunk 13]
```diff
@@ -622,3 +622,74 @@ def test_fetch_neurovault_ids():
...
+
+
+    with _TestTemporaryDirectory() as data_dir:
+        ### Second test case : a collection has been downloaded in the original version, user asks for resampled version
+
+        data_orig = neurovault.fetch_neurovault_ids(collection_ids=[4666], data_dir=data_dir,resample=False)
```
Comment (bthirion) [inline nilearn/datasets/tests/test_neurovault.py:661]:
```suggestion
        data_orig = neurovault.fetch_neurovault_ids(collection_ids=[4666], data_dir=data_dir, resample=False)
```

File: nilearn/datasets/tests/test_neurovault.py [hunk 14]
```diff
@@ -622,3 +622,74 @@ def test_fetch_neurovault_ids():
...
+        # Check that the original version is here
+        assert np.all([os.path.isfile(im_meta['absolute_path']) for im_meta in data_orig['images_meta']])
+
+        # Check that the resampled version is NOT here
+        assert not np.any([os.path.isfile(im_meta['resampled_absolute_path']) for im_meta in data_orig['images_meta']])
+
+        # Asks for the resampled version. This should only resample, not download. 
+        # (TODO - How can we test systematically that download does not happen ?  )
+        print("Asking for resampled version - Check here that download doesn't happen") 
+        data = neurovault.fetch_neurovault_ids(collection_ids=[4666], data_dir=data_dir,resample=True)
```
Comment (bthirion) [inline nilearn/datasets/tests/test_neurovault.py:672]:
```suggestion
        data = neurovault.fetch_neurovault_ids(collection_ids=[4666], data_dir=data_dir, resample=True)
```

File: nilearn/datasets/tests/test_neurovault.py [hunk 15]
```diff
@@ -622,3 +622,74 @@ def test_fetch_neurovault_ids():
...
+
+        # Check that the resampled version is here
+        assert np.all([os.path.isfile(im_meta['resampled_absolute_path']) for im_meta in data['images_meta']])
+
+        #And the original version should still be here as well
```
Comment (bthirion) [inline nilearn/datasets/tests/test_neurovault.py:677]:
```suggestion
        # And the original version should still be here as well
```

File: nilearn/datasets/tests/test_neurovault.py [hunk 16]
```diff
@@ -622,3 +622,74 @@ def test_fetch_neurovault_ids():
...
+        assert np.all([os.path.isfile(im_meta['absolute_path']) for im_meta in data['images_meta']])
+
+        # Load resampled images and check the affines
+
+        affines = [load_img(cur_im).affine for cur_im in data['images']]
+        assert np.all([np.all(affine==neurovault.STD_AFFINE) for affine in affines])
```
Comment (bthirion) [inline nilearn/datasets/tests/test_neurovault.py:683]:
```suggestion
        assert np.all([np.all(affine == neurovault.STD_AFFINE) for affine in affines])
```

File: nilearn/datasets/tests/test_neurovault.py [hunk 17]
```diff
@@ -622,3 +622,74 @@ def test_fetch_neurovault_ids():
...
+
+        # Check that the affines of the original version do not correspond to the resampled one
+        affines_orig = [load_img(cur_im).affine for cur_im in data_orig['images']]
+        assert not np.any([np.all(affine==neurovault.STD_AFFINE) for affine in affines_orig])
```
Comment (bthirion) [inline nilearn/datasets/tests/test_neurovault.py:687]:
```suggestion
        assert not np.any([np.all(affine == neurovault.STD_AFFINE) for affine in affines_orig])
```

File: nilearn/datasets/tests/test_neurovault.py [hunk 18]
```diff
@@ -622,3 +622,74 @@ def test_fetch_neurovault_ids():
...
+
+
+
+
+
+
+
```
Comment (bthirion) [inline nilearn/datasets/tests/test_neurovault.py:694]: Too many blank lines

File: nilearn/datasets/neurovault.py (hunk 9 above)
Comment (nicofarr) [reply nilearn/datasets/neurovault.py]: thanks, probably initially used 4x4x4 and switched to 3x3x3

**Comment by nicofarr** [comment]
> the way to do this would be to mock the fetch_file or simple_download function

I wasn't sure how to do, I suggest a solution using file access times and modification times. see commit d237782

**Comment by nicofarr** [comment]
I'd like to move on to write an example using this functionality. Initially the idea came when working on #1506 (will comment there).

**Comment by nicofarr** [comment]
I just found a backward compatibility issue: as I added a new field in image['id'] for the path to the resampled version, this field is not found for collections that were downloaded before the patch. I'll fix this.

**Comment by jeromedockes** [comment]
great! also, we are considering simplifying a bit the neurovault fetcher, when you have time could you have a look and give your opinion on #2283 ? (resampling functionality would be retained, of course)

File: nilearn/datasets/neurovault.py [hunk 19]
```diff
@@ -1507,11 +1513,42 @@ def _download_image_nii_file(image_info, collection, download_params):
...
+            image_url, tmp_path,
+            download_params['temp_dir'], verbose=download_params['verbose'])
+
+        # Resample here
+        print('Resampling...')
+        im_resampled = resample_img(img=tmp_path, target_affine=STD_AFFINE)
+        im_resampled.to_filename(resampled_image_absolute_path)
+
+        # Remove temporary file
+        os.remove(tmp_path)
```
Comment (jeromedockes) [inline nilearn/datasets/neurovault.py:1547]: could we maybe put the above in a try/finally block? to make sure the temp file is always removed

Comment (jeromedockes) [reply nilearn/datasets/neurovault.py:1547]: or if you prefer you can place the tempfile in `download_params['temp_dir']`. also, not sure you need the uuid, just appending something like "_not_resampled.nii.gz" should be fine. but using a uuid is also fine.

**Comment by jeromedockes** [comment]
> I just found a backward compatibility issue: as I added a new field in image['id'] for the path to the resampled version, this field is not found for collections that were downloaded before the patch. I'll fix this.

sorry could you elaborate a bit? `image[img_id]` is an int, right? do you mean that the metadata for some images will be missing the `resampled_absolute_path` and `resampled_relative_path` keys? but that would also be the case for images that are not resampled no?

**Comment by nicofarr** [comment]
> > I just found a backward compatibility issue: as I added a new field in image['id'] for the path to the resampled version, this field is not found for collections that were downloaded before the patch. I'll fix this.
> 
> sorry could you elaborate a bit? `image[img_id]` is an int, right? do you mean that the metadata for some images will be missing the `resampled_absolute_path` and `resampled_relative_path` keys? but that would also be the case for images that are not resampled no?

Yes sorry, indeed I meant the keys in the images metadata. This problem will always appear when a collection that has been downloaded before the patch.

**Comment by bthirion** [comment]
@RaphaelMeudec would you like to take a look at that one, and if possible, finalize it ?
Best,

**Comment by RaphaelMeudec** [comment]
Hi everyone! I would gladly help on this, @nicofarr do you still plan to work on it or can I help you finish and merge?

From what I see, what's left to do:

- [ ] Check backward compatibility with images downloaded before this resampling change (though, `image['id']` seems to [be already used](https://github.com/nilearn/nilearn/blob/29f64ef77013a8ae63c2458f0e2d75cf4007a062/nilearn/datasets/neurovault.py#L1651) in the neurovault fetcher?)
- [ ] Mock the downloads in the test
- [ ] Try/except the tmp_path remove

For now, we set the affine transformation and eventually switch to a configurable one if users ask for it. 

@jeromedockes do you see any other item to take care of?

**Comment by nicofarr** [comment]
Hi Raphael,

I would be happy if you could help ! I tend to not be able to take too much
time currently on this, except when there is a coding sprint... ;)

Thanks !

Nicolas

**Comment by RaphaelMeudec** [comment]
Hey everyone!

I'm having some issues on the testing side: the image downloaded by the neurovault mock is not an image but a json containing information related to the image, and therefore it is not nibabel compliant (error raised is `Cannot work out file type of "./neurovault/collection_993/image_5432.nii.gz"`).

In details, when calling for the download, the mock [returns a json object](https://github.com/nilearn/nilearn/blob/29f64ef77013a8ae63c2458f0e2d75cf4007a062/nilearn/datasets/tests/test_neurovault.py#L175) with the image information and not the image itself, which causes the non-compliance when [trying to load the downloaded image for the resampling](https://github.com/nicofarr/nilearn/blob/64ce5d2b658af66e93340fc6dcc9b86c5c92e771/nilearn/datasets/neurovault.py#L1541).

I see multiple options to this:

1. make the `_neurovault_one_image` return an actual image
2. mock the resampling call and keep the json in response

Option 1 seems the best to me, but it will extend the tests duration (don't know yet by how much).

Would love your opinion on this matter @nicofarr @GaelVaroquaux @jeromedockes ?

**Comment by jeromedockes** [comment]
> I'm having some issues on the testing side: the image downloaded by the neurovault mock is not an image but a json containing information related to the image, and therefore it is not nibabel compliant (error raised is `Cannot work out file type of "./neurovault/collection_993/image_5432.nii.gz"`).
>
> In details, when calling for the download, the mock [returns a json object](https://github.com/nilearn/nilearn/blob/29f64ef77013a8ae63c2458f0e2d75cf4007a062/nilearn/datasets/tests/test_neurovault.py#L175) with the image information and not the image itself, which causes the non-compliance when [trying to load the downloaded image for the resampling](https://github.com/nicofarr/nilearn/blob/64ce5d2b658af66e93340fc6dcc9b86c5c92e771/nilearn/datasets/neurovault.py#L1541).
>
> I see multiple options to this:
>
> 1. make the `_neurovault_one_image` return an actual image
> 2. mock the resampling call and keep the json in response
>
> Option 1 seems the best to me, but it will extend the tests duration (don't know yet by how much).

no the mock mimicks the actual neurovault api, this route does not
return an image but the json metadata for an image.
for example see the response for https://neurovault.org/api/images/2/

**Comment by RaphaelMeudec** [comment]
@jeromedockes Indeed! I narrowed the problem to the download of file being routed to `/media/images` (e.g `https://neurovault.org/media/images/4666/spmT_0002.nii.gz`). So the images handler does not differentiate `/api/images` and `/media/images`  (as [the handler matches based on the 2nd keyword](https://github.com/nilearn/nilearn/blob/29f64ef77013a8ae63c2458f0e2d75cf4007a062/nilearn/datasets/tests/test_neurovault.py#L206) of the route). Both requests are oriented to `_neurovault_one_image` even if they should have different outputs.

In light of this, I would recommend to create a handler to differentiate `/api/images` and `/media/images`:

- `/api/images` would keep the current behaviour
- `/media/images` would return a nii.gz file (which could be loaded from an asset in the tests folder)

**Comment by jeromedockes** [comment]
It seems to me that the request gets correctly routed to `_neurovault_file`
to see this, add and execute this test:
```
def test_mock_routing():
    res = requests.get("https://neurovault.org/api/media/images/4666/spmT_0002.nii.gz")
    print(res.content)
```

(and modify the return value of `_neurovault_file`)

however,

- as so far it wasn't used the `_neurovault_file` mock returns an empty string, 
- `_neurovault` dumps the result to json, which it should only do for
  the metadata and not for "media" urls

so my suggestion would be

- modify `_neurovault_file` to return a nifti image. as you said you
  could load an imaged shipped with nilearn, eg using
  `datasets.load_mni152_template`, but you can also generate a (small)
  fake one for example with
  `nilearn._utils.data_gen.generate_fake_fmri(length=1)[0]`. if you
  return a Nifti1Image the mocker will take care of serializing it to
  .nii.gz
- modify `_neurovault` so it doesn't convert to json when the url is
  "media"


to summarize the diff could look like:

```
--- a/nilearn/datasets/tests/test_neurovault.py
+++ b/nilearn/datasets/tests/test_neurovault.py
@@ -18,6 +18,7 @@ import pandas as pd
 import pytest
 
 from nilearn.datasets import neurovault
+from nilearn._utils.data_gen import generate_fake_fmri
 
 
 def _get_neurovault_data(random_seed=0):
@@ -177,7 +178,7 @@ def _neurovault_one_image(img_id):
 
 def _neurovault_file(parts, query):
     """Mocks the Neurovault API behind the `/api/media/` path."""
-    return ""
+    return generate_fake_fmri(length=1)[0]
 
 
 class _NumpyJsonEncoder(json.JSONEncoder):
@@ -205,6 +206,8 @@ def _neurovault(match, request):
     parts = list(filter(bool, info.path.split("/")))
     section = parts[1]
     result = handlers[section](parts[2:], _parse_query(info.query))
+    if section == "media":
+        return result
     return json.dumps(result, cls=_NumpyJsonEncoder).encode("UTF-8")
```

**Comment by RaphaelMeudec** [comment]
@jeromedockes not sure about the url you use: when download from neurovault, the url pattern seems to be `/media/images` and not `/api/media/images`. The collections/images we get from `_get_neurovault_data` have similar pattern. I've [modified the neurovault mock](https://github.com/nilearn/nilearn/pull/2696/commits/b04007d9b7fd1889df9f6d35ddd1ac2d8416dcf1) accordingly

I opened a new PR with the changes that is ready for review : #2696

**Comment by jeromedockes** [comment]
> @jeromedockes not sure about the url you use: when download from
> neurovault, the url pattern seems to be `/media/images` and not
> `/api/media/images`.
that's true, good point!

> The collections/images we get from `_get_neurovault_data` have similar pattern. I've [modified the neurovault mock](https://github.com/nilearn/nilearn/pull/2696/commits/b04007d9b7fd1889df9f6d35ddd1ac2d8416dcf1) accordingly
>
> I opened a new PR with the changes that is ready for review : #2696 

so the new PR replaces this one and we should close this one?

I'll review it

**Comment by RaphaelMeudec** [comment]
@jeromedockes Yes! I couldn't push to nicofarr's branch, hence the new PR. Thank you!

**Comment by jeromedockes** [comment]
closed in favor of #2696

**Pull request closed without merging**